manual/sphinx/developer_docs/data_types.rst

```diff
@@ -217,51 +217,201 @@ See the example ``examples/test-globalfield`` for more examples.
-syntax::
+If you wish to vectorise but can't use OpenMP then there is a serial
+verion of the macro:
+
+     BoutReal max=0.;
+     BOUT_FOR_SERIAL(i, region) {
+       max = f[i] > max ? f[i] : max;
+     }
+
+For loops inside parallel regions, there is ``BOUT_FOR_INNER``::
 
     Field3D f(0.0);
-    for (auto i : f.region(RGN_NOBNDRY)) {
-       f[i] = 1.0;
+    BOUT_OMP(parallel) {
+      BOUT_FOR_INNER(i, f.getMesh()->getRegion3D("RGN_ALL")) {
+         f[i] = a[i] + b[i];
+      }
+      ...
     }
+    
+If a more general OpenMP directive is needed, there is
+``BOUT_FOR_OMP``::
+
+  BoutReal result=0.;
+  BOUT_FOR_OMP(i, region, parallel for reduction(max:result)) {
+    result = f[i] > result ? f[i] : result;
+  }
+  
+The iterator provides access to the x, y, z indices::
 
-Available regions are:
+    Field3D f(0.0);
+    BOUT_FOR(i, f.getMesh()->getRegion3D("RGN_ALL")) {
+      f[i] = i.x() + i.y() + i.z();
+    }
 
--  `RGN_ALL`, which is the whole mesh;
+Note that calculating these indices involves some overhead: The
+iterator uses a single index internally, so integer division and
+modulo operators are needed to calculate individual indices.
+
+To perform finite difference or similar operators, index offsets can
+be calculated::
+
+  Field3D f = ...;
+  Field3D g(0.0);
+  BOUT_FOR(i, f.getMesh()->getRegion3D("RGN_NOBNDRY")) {
+    g[i] = f[i.xp()] - f[i.xm()];
+  }
+
+The ``xp()`` function produces an offset of ``+1`` in ``X``, ``xm()``
+an offset of ``-1`` in the ``X`` direction. ``xpp()``
+produces an offset of ``+2``, ``xmm()`` an offset of ``-2``, and
+similar functions exist for ``Y`` and ``Z`` directions. For other
+offsets there is a function ``offset(x,y,z)`` so that
+``i.offset(1,0,1)`` is the index at ``(x+1,y,z+1)``.
+
+Note that by default no bounds checking is performed. If the checking
+level is increased to 3 or above then bounds checks will be
+performed. This will have a significant (bad) impact on performance, so is
+just for debugging purposes. Configure with ``--enable-checks=3``
+option to do this.
+
+
+Tuning BOUT_FOR loops
+~~~~~~~~~~~~~~~~~~~~~
+
+The ``BOUT_FOR`` macros use two nested loops: The outer loop is OpenMP
+parallelised, and iterates over contiguous blocks::
+
+  BOUT_OMP(parallel for schedule(guided))
+  for (auto block = region.getBlocks().cbegin();
+       block < region.getBlocks().cend();
+       ++block)
+    for (auto index = block->first; index < block->second; ++index)
+
+The inner loop iterates over a contiguous range of indices, which
+enables it to be vectorised by GCC and Intel compilers.
+
+In order to both OpenMP parallelise, there must be enough blocks to
+keep all threads busy. In order to vectorise, each of these blocks
+must be larger than the processor vector width, preferably several
+times larger. This can be tuned by setting the maximum block size, set
+in ``include/bout/region.hxx``::
+
+  #define MAXREGIONBLOCKSIZE 64
```

Comment:
There's `mesh:maxregionblocksize` to change this at runtime -- the define just determines the default of this runtime parameter.